### PR TITLE
Handle statefulness in Cloud Monitoring exporter

### PR DESCRIPTION
**Pull request by cnnradams**
Copied from https://github.com/open-telemetry/opentelemetry-python/pull/898, also addressed @aabmass's comments

### CODE REVIEW

File: opentelemetry-exporter-cloud-monitoring/src/opentelemetry/exporter/cloud_monitoring/__init__.py [hunk 1]
```diff
@@ -217,10 +228,39 @@ def export(
...
+                    point.interval.start_time.nanos = int(
+                        self._exporter_start_time_nanos
+                    )
+                else:
+                    # The aggregation reset the last time it was exported
+                    (
+                        point.interval.start_time.seconds,
+                        point.interval.start_time.nanos,
+                    ) = map(
+                        int, divmod(self._last_updated[updated_key] + 1e6, 1e9)
```
Comment (aabmass) [inline opentelemetry-exporter-cloud-monitoring/src/opentelemetry/exporter/cloud_monitoring/__init__.py:257]: Would be good to have a comment and/or link to those docs here about the 1ms.

File: opentelemetry-exporter-cloud-monitoring/src/opentelemetry/exporter/cloud_monitoring/__init__.py [hunk 2]
```diff
@@ -1,3 +1,5 @@
+# pylint: disable=too-many-locals
```
Comment (AndrewAXue) [inline opentelemetry-exporter-cloud-monitoring/src/opentelemetry/exporter/cloud_monitoring/__init__.py:1]: Why is this needed? And can you resolve this without resorting to disabling tests?

File: opentelemetry-exporter-cloud-monitoring/tests/test_cloud_monitoring.py [hunk 3]
```diff
@@ -1,3 +1,4 @@
+# pylint: disable=too-many-statements
```
Comment (AndrewAXue) [inline opentelemetry-exporter-cloud-monitoring/tests/test_cloud_monitoring.py:1]: Why is this needed? And can you resolve this without resorting to disabling tests?

File: opentelemetry-exporter-cloud-monitoring/src/opentelemetry/exporter/cloud_monitoring/__init__.py [hunk 4]
```diff
@@ -217,10 +228,39 @@ def export(
...
+
+            if (
+                metric_descriptor.metric_kind
+                == MetricDescriptor.MetricKind.CUMULATIVE
+            ):
+                if (
+                    instrument.meter.batcher.stateful
+                    or updated_key not in self._last_updated
+                ):
+                    # The aggregation has not reset since the exporter has started up, so that is the start time
```
Comment (AndrewAXue) [inline opentelemetry-exporter-cloud-monitoring/src/opentelemetry/exporter/cloud_monitoring/__init__.py:244]: nit: Line too long

File: opentelemetry-exporter-cloud-monitoring/src/opentelemetry/exporter/cloud_monitoring/__init__.py [hunk 5]
```diff
@@ -217,10 +228,39 @@ def export(
...
             updated_key = (metric_descriptor.type, record.labels)
-            last_updated_seconds = self._last_updated.get(updated_key, 0)
-            if seconds <= last_updated_seconds + WRITE_INTERVAL:
+            last_updated_time = self._last_updated.get(updated_key, 0)
+            last_updated_time_seconds = last_updated_time // 1e9
+            if seconds <= last_updated_time_seconds + WRITE_INTERVAL:
                 continue
-            self._last_updated[updated_key] = seconds
+
+            if (
```
Comment (AndrewAXue) [inline opentelemetry-exporter-cloud-monitoring/src/opentelemetry/exporter/cloud_monitoring/__init__.py:236]: Now thats theres quite a bit more logic about the timeSeries Point creation, can you move that out to  seperate function?

File: opentelemetry-exporter-cloud-monitoring/src/opentelemetry/exporter/cloud_monitoring/__init__.py (hunk 2 above)
Comment (cnnradams) [reply opentelemetry-exporter-cloud-monitoring/src/opentelemetry/exporter/cloud_monitoring/__init__.py]: removed it by moving to new fn

File: opentelemetry-exporter-cloud-monitoring/tests/test_cloud_monitoring.py (hunk 3 above)
Comment (cnnradams) [reply opentelemetry-exporter-cloud-monitoring/tests/test_cloud_monitoring.py:1]: not sure honestly, but I wouldn't consider this removing tests, more like removing a lint exception that I don't believe matters

File: opentelemetry-exporter-cloud-monitoring/src/opentelemetry/exporter/cloud_monitoring/__init__.py (hunk 4 above)
Comment (cnnradams) [reply opentelemetry-exporter-cloud-monitoring/src/opentelemetry/exporter/cloud_monitoring/__init__.py]: fixed

File: opentelemetry-exporter-cloud-monitoring/src/opentelemetry/exporter/cloud_monitoring/__init__.py (hunk 5 above)
Comment (cnnradams) [reply opentelemetry-exporter-cloud-monitoring/src/opentelemetry/exporter/cloud_monitoring/__init__.py]: moved

File: opentelemetry-exporter-cloud-monitoring/src/opentelemetry/exporter/cloud_monitoring/__init__.py (hunk 1 above)
Comment (cnnradams) [reply opentelemetry-exporter-cloud-monitoring/src/opentelemetry/exporter/cloud_monitoring/__init__.py]: added

File: opentelemetry-exporter-cloud-monitoring/src/opentelemetry/exporter/cloud_monitoring/__init__.py [hunk 6]
```diff
@@ -65,6 +69,11 @@ def __init__(
                 random.randint(0, 16 ** 8)
             )
 
+        (
+            self._exporter_start_time_seconds,
+            self._exporter_start_time_nanos,
+        ) = divmod(time_ns(), 1e9)
```
[the remark this answers is not in the document]
Comment (cnnradams) [reply opentelemetry-exporter-cloud-monitoring/src/opentelemetry/exporter/cloud_monitoring/__init__.py]: good suggestion, no longer using all of the int()s

File: opentelemetry-exporter-cloud-monitoring/src/opentelemetry/exporter/cloud_monitoring/__init__.py [hunk 7]
```diff
@@ -154,8 +163,10 @@ def _get_metric_descriptor(
...
 
-        if isinstance(record.aggregator, SumAggregator):
-            descriptor["metric_kind"] = MetricDescriptor.MetricKind.GAUGE
+        if isinstance(record.aggregator, SumAggregator) and not isinstance(
+            record.instrument, UpDownCounter
+        ):
+            descriptor["metric_kind"] = MetricDescriptor.MetricKind.CUMULATIVE
         else:
             logger.warning(
                 "Unsupported aggregation type %s, ignoring it",
```
[the remark this answers is not in the document]
Comment (aabmass) [reply opentelemetry-exporter-cloud-monitoring/src/opentelemetry/exporter/cloud_monitoring/__init__.py:173]: I meant the logger warning message, since SumAggregator is supported but UpDownCounter is not (for now)

[the remark this answers is not in the document]
Comment (cnnradams) [reply opentelemetry-exporter-cloud-monitoring/src/opentelemetry/exporter/cloud_monitoring/__init__.py]: updated
